keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,370 @@
-At least the following installation scenarios should be supported:
-  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
-  2. A user should be able to install a specific `appVersion` of an operator
 
-### Support multiple API versions in the package repository
+## Proposals
+
+### With regards to package version and operator version
+
+Two proposals: singular and composite package/operator version. In the case of potential operator "bundles" (operators containing multiple app versions) both would end up "hiding" those app versions behind the package/operator version.
+
+#### Singular package/operator version (largely, the current state of things)
+
+##### Pros
+
+- Visually simple
+- It can be known just from the package version whether or not it contains a certain operator-related feature or bug fix
+
+##### Cons
+
+- Operator users care about the application version, not a "synthetic" version
+- Makes it complex to have an "app version"-based backport strategy. Customers might not be able to use operator based on newer app versions due to constraints outside of our control, but would still want to receive new operator-related features and bug fixes
+- Continuing from the second "pro", it might be possible that subsequent package versions with different app versions don't contain an operator-related feature or bug fix that only makes sense for a specific app version (e.g., version N+1 based on app version M doesn't include an operator feature available on version N based on app version M+1 because the operator feature only makes sense for app version M)
+- The progression of the package version is confusing with regards to the application version progression
+
+  Example:
+
+  | Time | Package/operator version | App version        | Comment                                            |
+  | ---- | ---------------          | ------------------ | -------------------------------------------------- |
+  | T0   | 1.0.0                    | 2.3.0              | Initial release based on app version 2.3.x         |
+  | T1   | 2.0.0                    | 3.0.0              | Release based on app version 3.0.x                 |
+  | T2   | 2.1.0                    | 2.3.0              | New operator-related feature for app version 2.3.x |
+  | T2   | 2.2.0                    | 3.0.0              | New operator-related feature for app version 3.0.x |
+  | T3   | 2.3.0                    | 2.3.1              | Bug fix release for app version 2.3.x              |
+
+  The example above illustrates two sources of confusion:
+  - A higher package version can contain a lower app version
```

Comment:
This doesn't have to be true. We should maintain the release branch and backport the features if necessary. 
Please read how kubernetes maintains a  release branch and backports the features that are necessary. 
The development is done in `master` and in case if operator developer or users want a feature to be backported. Its backported to the old release branch. 
```
 | Time | Package/operator version | App version        | Comment                                            |
  | ---- | ---------------          | ------------------ | -------------------------------------------------- |
  | T0   | 1.0.0                    | 2.3.0              | Initial release based on app version 2.3.x         |
  | T1   | 2.0.0                    | 3.0.0              | Release based on app version 3.0.x                 |
  | T2   | 1.1.0                    | 2.3.0              | New operator-related feature for app version 2.3.x |
  | T2   | 2.2.0                   | 3.0.0              | New operator-related feature for app version 3.0.x |
  | T3   | 1.1.1                     | 2.3.1              | Bug fix release for app version 2.3.x              |
```

Reply:
If I understood this correctly, in that case we'd be *statically reserving* package/operator versions for app version families (e.g., 2.3, 3.0), right? This is something we talked about last week and we should add this to the KEP.

In the case though (on your example above), what happens when there's a breaking change in version `1.1.0` (based on app version `2.3.0`)? Given that `2.2.0` exists (based on app version `3.0.0`. Does the package/operator version go to `3.0.0` (based on app version `2.3.0`)?

I'd also argue that statically reserving versions breaks SemVer.

Reply:
> In the case though (on your example above), what happens when there's a breaking change in version 1.1.0 (based on app version 2.3.0)? Given that 2.2.0 exists (based on app version 3.0.0. Does the package/operator version go to 3.0.0 (based on app version 2.3.0)?

this is independent of versioning, if there is a breaking change in base tech based approach, the same will happen here. 

> I'd also argue that statically reserving versions breaks SemVer. 
we start from a version and backport features or patch/minor updates in base tech. This is compatible with semver

Reply:
from the above I mean, we shouldn't expect the versioning to save us from breaking changes or make it easier for breaking changes. It needs to be tackled independently of which versioning scheme we use.

Reply:
By reserving operator version ranges you're implicitly connecting operator version to the app version. End-users will have to keep the version matrix in their heads. I'd rather do it explicitly.

Reply:
the operator itself is `base-tech + logic to operate the base-tech` so having that connected to `base-tech` and `logic to operate the app` feels natural to me.

Reply:
and end-users in both cases can install using `--app-version=2.3.0` flag and not having to know the version of operator